# AgoRBNS Analysis Pipeline

This repository contains analysis code for two publications:

1. McGeary SE, et al. "The biochemical basis of microRNA targeting efficacy." Science. 2019.
   [Link to paper](https://www.science.org/doi/10.1126/science.aav1741)

2. McGeary SE, et al. "MicroRNA 3'-compensatory pairing occurs through two binding modes, with affinity shaped by nucleotide identity and position." eLife. 2022.
   [Link to paper](https://elifesciences.org/articles/73188)

## Data Access

Raw sequencing data is available from GEO under the following accession numbers:
- GSE140220 (Science 2019 paper)
- GSE196458 (eLife 2022 paper)

## Requirements

- Conda (Miniconda or Anaconda) (tested using Miniconda 23.10)
- Git

## Installation and Setup

1. Clone this repository:
```bash
    git clone https://github.com/smcgeary/agorbns.git
    cd AGORBNS
```

2. Create the agorbns_sratools conda environment:
```bash
    conda env create -f environment_agorbns_sratools.yml
```

3. Create the agorbns conda environment:
```bash
    conda env create -f environment_agorbns.yml
```

3. Activate the conda environment:
```bash
    conda activate agorbns
```

4. Verify the environment is active:
```bash
   conda env list
```
   The active environment (agorbns) should be marked with *

5. When finished working, deactivate the environment:
```bash
    conda deactivate
```

Note: If the environment.yml file is updated, you can update your environment with:
```bash
    conda env update -f environment.yml
```

## Data acquisition

To download all data associated with both papers:

1. Activate the agorbns_sratools conda environment:
```bash
    source activate agorbns_sratools
```


2. Assemble the metadata-table required for parsing the data:
```bash
    python SRA_download_scripts/make_SRA_and_sample_table.py
```

2. Next download each dataset:
```bash
    bash SRA_download_scripts/download_SRA_files.sh
```

1. Deactivate the agorbns_sratools conda environment:
```bash
    conda deactivate
```


## Repository Structure

Analyses proceed through three primary stages. First, the read data is extracted
using the functionality in the PreProcessReads directory. Then site types are
assigned to reads using the AssignSiteTypes directory. Then, biochemical
modeling that enables Kd estimation is performed using functionality in the
SolveForKds directory.


## Usage

The basic commands to perform from the command line are listed in order leading
up to each Figure sub-panel in either the `general/Figures2019.R` or 
`general/Figures2022.R` script.

These are generated by entering R interactively from the terminal:
```bash
    R
```

and loading it by calling:
```R
    source("general/Figures2019.R")
```
or
```R
    source("general/Figures2022.R")
```


Navigating to `MakeFigure1()` function in `general/Figures2019.R" shows the
code:
```R
## To make this figure need run these commands in the command line:
## make PreprocessAllEquilibrium
## make AssignSites mirna=miR-1 exp=equilibrium n_constant=5 buffer=1 sitelist=canonical
## python SolveForKds/MakeSiteCountTable.py miR-1 equilibrium 5 canonical -buffer
message("Making Fig. 1")
PlotEquilSiteWithInput("miR-1", 7, sitelist="canonical", combined=FALSE,
                        buffer=TRUE, pdf.plot="1.C")
message("Done 1.C")
## python SolveForKds/MakeMultiSiteCountTable.py miR-1 equilibrium 5 canonical -buffer3p
## Rscript SolveForKds/FitSiteKds.R miR-1 equilibrium 5 canonical -buffer3p -nocombI -single
PlotSiteEnrichments("miR-1", sitelist="canonical", combined=FALSE,
                    buffer=TRUE, write_kds=TRUE, pdf.plot="1.D")
message("Done 1.D")
## make AssignSites mirna=miR-1 exp=equilibrium n_constant=5 buffer=1 sitelist=resubmissionfinal
## python SolveForKds/MakeSiteCountTable.py miR-1 equilibrium 5 resubmissionfinal -buffer
## python SolveForKds/MakeMultiSiteCountTable.py miR-1 equilibrium 5 resubmissionfinal -buffer
## Rscript SolveForKds/FitSiteKds.R miR-1 equilibrium 5 resubmissionfinal -buffer -nocombI -single
PlotSiteEnrichments("miR-1", combined=FALSE, remove_sites=FALSE, buffer=TRUE,
                    pdf.plot="1.E")
message("Done 1.E")
```

There are three commands that must be completed in order to make Figure 1.C:

```bash
make PreprocessAllEquilibrium
make AssignSites mirna=miR-1 exp=equilibrium n_constant=5 buffer=1 sitelist=canonical
python SolveForKds/MakeSiteCountTable.py miR-1 equilibrium 5 canonical -buffer
```

Each of these commands must be fully completed before progressing to the next
one. The makefile is currently configured to use the SLURM job submission
system on the O2 computing cluster associated with the Harvard Medical School
research campus. The makefile would need to be modified to be used with other
submission systems, such as IBM LSF.


## Citation

If you use this code in your research, please cite:

McGeary, S.E.\*, Lin, K.S.\*, Shi, C.Y., Pham, T.M., Bisaria, N., Kelley, G.M., and Bartel, D.P. (2019). The biochemical basis of miRNA targeting efficacy. Science, 366, eaav1741.

McGeary, S.E.\*, Bisaria, N.\*, Pham, T.M., Wang, P.Y., and Bartel, D.P. (2022). MicroRNA 3â€²-compensatory pairing occurs through two binding modes, with affinity shaped by nucleotide identity and position. eLife, 11, e69803.